### ISSUE TITLE
Add multi-user support for Spark Notebook

### ISSUE THREAD

**Issue by ogirardot** [andypetrella/spark-notebook#232]

**Comment by andypetrella** [comment]
Just a bit of info before the description is gracefully filled in by @DaFellow (dunno his github account... yet ^^).

The notebook server will start one (and only one) session at a time for each notebook started.
A session is 1o1 mapped with a java process[1].

So, the thing is that there is no notion of users, and even though there is one, I like to keep the door to live collaboration opened (at least) (see also #162 ), that means that it'll be something we'll have to think seriously ;-).

Regarding pure authentication there is already #193 that I'd love landing soon (0.6.0).

So now, since multi-user can have different meaning, I'll let you guyz drop in your thoughts at your pace.
(note: this is some info that are not there to influence your wishes!)

[1] I'm thinking at creating a trait that will allow these processes to be scheduled/started elsewhere too (on marathon for instance)

**Comment by ogirardot** [comment]
great infos ! Thanks !

**Comment by YannBrrd** [comment]
Hi,

Here are the things I have in mind : 
- provide an authentication form to log in to spark notebook server (based on PAM ?)
- have Spark Notebook using the user grants to access file system (or any other source, like HDFS for exemple) using impersonation
- on top of that, keep collaboration between users, keeping the security grants...

**Comment by ekovacs** [comment]
it would be awesome to send tasks to mesos jobs/resource  managers / schedulers like:
- Singularity
- Chronos
- Marathon
- Aurora
- Spark-jobserver

**Comment by andypetrella** [comment]
@ekovacs I have this idea for a while now, and I think it's a must have. But also deserves its own ticket.
The basic idea would be to create a middle trait between the spark-notebook and the code that forks the process + notify where the remote actor is listening. 

So that the notebook could spawn processes wherever the implementor is able to run a spark repl + akka remote

@YannBrrd I see, I'm just wondering how the credentials would be passed from the main server to the "remote" process. I do think that at least a login mechanism is very needed, after the user is connected he is allowed to run any code within a notebook, including authentication to the cluster maybe.

#234 might also help, because if it would be possible to create some var env that will be applied at the notebook level only (one by one) this would allow us to create generic authentication procedure, like JAAS (for kerberos for instance) and waiting for the right system properties pointing to the user's credentials.

I miss the big picture on what we would like to have, I'd love to see that one, and helping at its creation (if I would be needed of course).
A small diagram maybe?

**Comment by YannBrrd** [comment]
Hi Andy,

Heads up on this topic.

Global idea is : 
1. Centralized server, with Spark Notebook installed by Ops, with configuration done
2. Web interface for end users to login
3. A process to launch a Spark Notebook instance on the user behalf (and yet still under the same server)

Global idea is what is implemented by [Jupyterhub ](https://github.com/jupyter/jupyterhub/) for Jupyter.

Wondering if a fork and adapt for Spark-notebook could do the trick (I guess it could). And then adapt various spawners ([sudospawner ](https://github.com/jupyter/sudospawner) & [docker spawner ](https://github.com/jupyter/dockerspawner) would be a good start I guess).

Would allow to deploy centrally a Spark-Notebook As A Service in a secured environment.

Will investigate further in the coming weeks.

If diagram needed, let me know... 

:smile: 

Yann

**Comment by andypetrella** [comment]
Hey @YannBrrd, interesting idea.

I've the same thoughts, but using Marathon (when on Mesos) as an external controller.
This spawner would be like a generic interface to communicate with this kind of manager maybe?

**Comment by YannBrrd** [comment]
Right. I'm thinking about building a generic UI, based on LDAP maybe + plugins to spawn processes (sudo, docker, marathon, ...). 

Looking at nodejs as a stack ATM, intead of forking Jupytherhubv (still keeping the principles)... Depends on the time I have ahead. 

Regarding Marathon, I'm not yet aware of how this would work. I'll need an update on this...

If I find some time, I'll try to build something in the coming weeks.

**Comment by andypetrella** [comment]
Please don't use nodejs. Let's have a simple stack with simple deployment.

Try to stick with JVM based technologies, for instance Play will probably offer everything this use case will need, which is fair simple so far.

**Comment by YannBrrd** [comment]
Ok

Le sam. 6 févr. 2016 13:08, Andy Petrella <notifications@github.com> a
écrit :

> Please don't use nodejs. Let's have a simple stack with simple deployment.
>
> Try to stick with JVM based technologies, for instance Play will probably
> offer everything this use case will need, which is fair simple so far.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/andypetrella/spark-notebook/issues/232#issuecomment-180747017>
> .
>
-- 
Cordialement,
Yann Barraud

**Comment by YannBrrd** [comment]
Been investigating a bit. Sticking to the JVM will require to issue a system call to launch process... Keeping thinking about design...

**Comment by YannBrrd** [comment]
@andypetrella Marathon is certainly a good idea, but it would restrict usage of Spark Notebook Spawner to Mesos users, which is not a good idea from my point of view. Marathon is one of various options we should have, like : 

- Basic OS based spawner 
- Docker spawner
- Marathon spawner (+docker on top also)
